substrate/client/network/sync/src/engine.rs

```diff
@@ -245,6 +245,9 @@ pub struct SyncingEngine<B: BlockT, Client> {
 	/// The `PeerId`'s of all boot nodes.
 	boot_node_ids: HashSet<PeerId>,
 
+	/// A channel to get target block header if we skip over proofs downloading during warp sync.
+	warp_sync_target_block_header_rx: Option<oneshot::Receiver<<B as BlockT>::Header>>,
+
```

Comment:
I think this is fine for now but in the future there should be some kind of syncing strategy selection like:

```rust
enum Strategy {
    Warp(WarpSync),
    Full(FullSync),
    Whatever(...)
}

enum Action {
    SendRequest(Vec<u8>),
    SendResponse(Vec<u8>),
    ImportBlock(Block),
}

enum SyncEvent {
    ResponseReceived(Vec<u8>),
}

trait Strategy {
     fn register_event(&mut self, event: SyncEvent);
     fn next_action(&mut self) -> Option<SyncAction>;
}

impl SyncingEngine {
    async fn run(mut self) {
        loop {
            tokio::select! {
                response = self.pending_responses.select_next_some() => {
                    self.strategy.register_event(SyncEvent::ResponseReceived(response));
                }
            }

            if let Some(action) = self.strategy.next_action() {
                // TODO: perform action
            }
        }
    }
}
```

Details of the syncing algorithm would be opaque to `SyncingEngine` and it would only deal with I/O and dispatch the events to the correct syncing strategy which would deal with the logic of that particular strategy and emit I/O-related events to `SyncingEngine`. How feasible this is to remains to be seen but worth looking into.

Reply:
Yeah that is a good idea. Goes into the direction of: https://github.com/paritytech/polkadot-sdk/issues/532

Reply:
@altonen Thanks for the snipped provided! My idea was to gradually move all polling to the upper levels (`SyncingEngine`) first, otherwise it's difficult to do everything in one go. And `WarpSync` requiring polling was quite an annoying part, so I started from it.

Reply:
Yes I agree this shouldn't be done in this PR and I think your approach of first gradually removing stuff away from `ChainSync` is good but I just wanted to highlight what the ultimate plan is.